torch/csrc/autograd/engine.cpp

```diff
@@ -863,15 +863,33 @@ auto Engine::execute(const edge_list& roots,
       /* depth */ not_reentrant_backward_call ? 0 : total_depth + 1,
       /* cpu_ready_queue */ local_ready_queue);
 
+  // If we receive a single root, skip creating extra root node
+  bool skip_dummy_node = roots.size() == 1;
+  auto graph_root = skip_dummy_node ?
+    roots.at(0).function :
+    std::make_shared<GraphRoot>(roots, inputs);
+
   // Now compute the dependencies for all executable functions and queue the root
-  auto graph_root = std::make_shared<GraphRoot>(roots, inputs);
   compute_dependencies(graph_root.get(), *graph_task);
 
   if (!outputs.empty()) {
     graph_task->init_to_execute(*graph_root, outputs, accumulate_grad);
   }
 
-  execute_with_graph_task(graph_task, graph_root);
+  if (skip_dummy_node) {
+    auto inputs_ = std::vector<Variable>();
+    inputs_.reserve(graph_root->num_inputs());
+    for (size_t i = 0; i < graph_root->num_inputs(); ++i) {
```

Comment:
I guess this is inspired by the GraphRoot class constructor? It is doing some non-trivial stuff so it might be nice to put that in a helper function if possible to make sure they don't diverge in the future.

Reply:
This part handles the mismatch between node->num_inputs() and inputs.size() (which seems to be always 1). The code that handles this in the original code is split between evaluate_function and input_buffer.add. First, in evaluate_function, a new vector is created with size, num_inputs() (which may be greater than inputs.size()). The output of the fn.apply is added to the ith position of the input_buffer, where i is determined by input_nr

Reply:
Ok.